src/wire/ipv6routing.rs

```diff
@@ -0,0 +1,658 @@
+        }
+    }
+}
+
+/// Setter methods for the RPL Source Routing Header routing type.
+impl<T: AsRef<[u8]> + AsMut<[u8]>> Header<T> {
+    /// Set the Ipv6 Home Address
+    ///
+    /// # Panics
+    /// This function may panic if this header is not the Type 2 Routing Header routing type.
+    pub fn set_home_address(&mut self, value: Address) {
+        let data = self.buffer.as_mut();
+        data[field::HOME_ADDRESS].copy_from_slice(value.as_bytes());
+    }
+}
+
+/// Setter methods for the RPL Source Routing Header routing type.
+impl<T: AsRef<[u8]> + AsMut<[u8]>> Header<T> {
+    /// Set the number of prefix octects elided from addresses[1..n-1].
+    ///
+    /// # Panics
+    /// This function may panic if this header is not the RPL Source Routing Header routing type.
+    pub fn set_cmpr_i(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        let raw = (value << 4) | (data[field::CMPR] & 0xF);
+        data[field::CMPR] = raw;
+    }
+
+    /// Set the number of prefix octects elided from the last address (addresses[n]).
+    ///
+    /// # Panics
+    /// This function may panic if this header is not the RPL Source Routing Header routing type.
+    pub fn set_cmpr_e(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        let raw = (value & 0xF) | (data[field::CMPR] & 0xF0);
+        data[field::CMPR] = raw;
+    }
+
+    /// Set the number of octects used for padding after addresses[n].
+    ///
+    /// # Panics
+    /// This function may panic if this header is not the RPL Source Routing Header routing type.
+    pub fn set_pad(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        data[field::PAD] = value << 4;
+    }
+
+    /// Set address data
+    ///
+    /// # Panics
+    /// This function may panic if this header is not the RPL Source Routing Header routing type.
+    pub fn set_addresses(&mut self, value: &[u8]) {
+        let data = self.buffer.as_mut();
+        let len = data[field::LENGTH];
+        let addresses = &mut data[field::ADDRESSES(len)];
+        addresses.copy_from_slice(value);
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> fmt::Display for Header<&'a T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match Repr::parse(self) {
+            Ok(repr) => write!(f, "{}", repr),
+            Err(err) => {
+                write!(f, "IPv6 Routing ({})", err)?;
+                Ok(())
+            }
+        }
+    }
+}
+
+/// A high-level representation of an IPv6 Routing Header.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Repr<'a> {
+    Type2 {
+        /// The type of header immediately following the Routing header.
+        next_header:    Protocol,
+        /// Length of the Routing header in 8-octet units, not including the first 8 octets.
+        length:         u8,
+        /// Number of route segments remaining.
+        segments_left:  u8,
```

Comment:
This repr doesn't need `length` or `segments_left` they will always be 2 and 1 respectively. It is an invalid packet if they are not.

Reply:
I went back and forth on this. I decided to leave them explicit as I imagined that downstream code would not want these values implicit. If I remove those fields, should I provide methods on the Repr that return the number of length and number of segments left?

I have thought about this validity problem a lot, but thought validation would be handled outside of the wire modules. I have not validated the values in the Repr (for fragmentation, hop by hop and now routing) are all valid with respect to each other. For example, I do not check that the `cmpr_i` value is valid based on the vector of addresses being set.

Reply:
> I decided to leave them explicit as I imagined that downstream code would not want these values implicit.

I suppose it could cut down on magic numbers in downstream code when parsing a Type 2 routing header. No harm in leaving it in for now. We can always create an issue to discuss it further.

> I have thought about this validity problem a lot, but thought validation would be handled outside of the wire modules.

I'm not suggesting validating the packet. I was just enforcing that we know a Type 2 headers effective length is 2.